src/providers/baremetal.rs

```diff
@@ -171,26 +170,27 @@ pub struct Machine {
 impl super::Launcher for Machine {
     type MachineDescriptor = Setup;
 
+    #[instrument(level = "debug", skip(self))]
     fn launch<'l>(
         &'l mut self,
         l: super::LaunchDescriptor<Self::MachineDescriptor>,
-    ) -> Pin<Box<dyn Future<Output = Result<(), Error>> + Send + 'l>> {
+    ) -> Pin<Box<dyn Future<Output = Result<(), Report>> + Send + 'l>> {
         Box::pin(async move {
-            self.log = Some(l.log);
-            let log = self.log.as_ref().expect("Baremetal machine uninitialized");
-
             let mut dscs = l.machines.into_iter();
             let (name, mut setup) = dscs
                 .next()
-                .ok_or_else(|| format_err!("Cannot initialize zero machines"))?;
+                .ok_or_else(|| eyre!("Cannot initialize zero machines"))?;
             for (discarded_name, discarded_setup) in dscs {
-                warn!(log, "Discarding duplicate connections to same machine";
-                    "name" => &discarded_name,
-                    "addr" => &discarded_setup.addr[0],
+                tracing::warn!(
+                    name = %discarded_name,
+                    addr = %discarded_setup.addr[0],
+                    "Discarding duplicate connections to same machine",
                 );
             }
 
-            let addr = try_addrs(&mut setup, &log, l.max_wait).await?;
+            let addr = try_addrs(&mut setup, l.max_wait)
+                .await
+                .wrap_err("probing baremetal addresses")?;
```

Comment:
similar to above, this seems like something that belongs in a span rather or needs to be rewritten to represent an error.

Reply:
Only tangentially related, but the code this ends up calling (`try_addrs`) actually accumulates _multiple_ errors, and I'm not sure how to nicely collapse that into one error. What I'm now doing is
https://github.com/jonhoo/tsunami/blob/1e0dadc760890e90d9596cfb575dae9b73ccd298/src/providers/baremetal.rs#L170-L177

but that feels pretty dirty... Ideas?

Reply:
aah, yes, I definitely have thoughts about this. The error trait doesn't really allow for representing errors as anything other than a singly linked list.

Using `wrap_err` here is probably something you definitely don't want to do, because it will use the `Display` method for the errors which will only show the first error message in that report's chain of errors.

Ideally, I'd say don't try to represent errors past this point as errors anymore. Just stick with the language of error reports and return a Vec<Report> if possible.

That said, ideally there would be a way to construct an error with multiple source errors. I have an [RFC](https://github.com/yaahc/rfcs/blob/master/text/0000-dyn-error-generic-member-access.md#example-use-cases-this-enables) in progress that actually makes this possible. Specifically, check out the 3rd bullet in list i linked to about representing source errors as a tree of errors. If that RFC ever merges I'll update eyre and company to be able to wrap single errors or sets of errors, and correctly traverse the tree of errors when reporting.

Reply:
Hmm, a `Vec<Report>` seems pretty unwieldy though. Unless I change all error types in the application to `Vec<Report>` to cater to this specific cast. I think defaulting to just displaying the first is actually perfectly fine in this _particular_ case. The error here is if a particular host has multiple IP addresses, and we fail to connect to _all_ of them. Reporting the error just for the first IP address in the common case seems fine to me. Thanks for the RFC link though, will definitely be watching that!